std/typecons.d

```diff
@@ -6896,3 +6897,503 @@ unittest
+        return value;
+    }
+
+    /**
+     * If there is a held value, return it; otherwise return other.
+     * Params: other = value to return if this instance is empty
+     * Bugs: These functions are currently not `nothrow` or `@nogc`. See
+     * $(BUGZILLA 12647).
+     * See_Also: $(REF Algebraic, std, variant)
+     */
+    // TODO: Do lazy parameters offer ref yet?
+    auto ref CommonType!(inout(T), U) getOrElse(U : T)(lazy U other) inout
+    {
+        return isDefined ? value : other;
+    }
+
+    /// Ditto
+    Option!(CommonType!(T, U)) orElse(U)(lazy Option!U other)
+        if (!is(CommonType!(T, U) == void))
+    {
+        alias Super = CommonType!(T, U);
+        return isDefined ? some!Super(this.value) : other.upcast!Super;
+    }
+
+    version(StdDdoc)
+    {
+        /// Ditto
+        Algebraic!(inout(T), U) or(lazy U other) inout;
+    }
+    else
+    {
+        template or(U)
+        {
+            import std.variant : Algebraic;
+            Algebraic!(T, U) or(lazy U other)
+            {
+                return isDefined ? typeof(return)(value) : typeof(return)(other);
+            }
+
+            Algebraic!(const T, U) or(lazy U other) const
+            {
+                return isDefined ? typeof(return)(value) : typeof(return)(other);
+            }
+
+            Algebraic!(immutable T, U) or(lazy U other) immutable
+            {
+                return isDefined ? typeof(return)(value) : typeof(return)(other);
+            }
+        }
+    }
+}
+
+/// Convert from null to Option as soon as possible when interfacing with code that uses null:
+nothrow @safe @nogc unittest
+{
+    import std.algorithm.iteration : map;
+    import std.range : only;
+
+    static int* getX(bool yes) nothrow @safe @nogc
+    {
+        static int x = 42;
+        return yes? &x : null;
+    }
+
+    auto values = only(false, true, false).map!getX.map!option;
+    assert(values[0].empty);
+    assert(!values[1].empty && *values[1].front == 42);
+    assert(values[2].empty);
+}
+
+/// Use $(REF joiner, std, algorithm, iteration) to $(I flatten) a range of `Option`s:
+nothrow @safe @nogc unittest
+{
+    import std.algorithm.comparison : equal;
+    import std.algorithm.iteration : joiner;
+    import std.range : only;
+    static maybeValues = [some(42), Option!int(none), some(24)];
+    assert(maybeValues.joiner.equal(only(42, 24)));
+}
+
+/// There is no `flatMap`; use `joiner` with $(REF map, std, algorithm, iteration):
+@safe unittest
+{
+    static import std.uni;
+    import std.algorithm.comparison : equal;
+    import std.algorithm.iteration : joiner, map;
+
+    static maybeValues = [Option!string(none), some("hello"), some("world")];
+    assert(maybeValues.joiner.map!(std.uni.toUpper).joiner(" ").equal("HELLO WORLD"));
+}
+
+/// Use `foreach` or $(REF each, std, algorithm, iteration) with `joiner` to perform an operation with side effects on `Option`s:
```

Comment:
``` `…` ``` is the same as `$(D …)`.

`REF` is from D-Programming-Language/dlang.org/pull/1184.